File: app/src/main/java/com/rubyhuntersky/angleedit/app/MainActivity.java
package com.rubyhuntersky.angleedit.app;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.v7.app.ActionBarActivity;
import android.util.Log;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;

public class MainActivity extends ActionBarActivity implements XmlDocumentFragment.XmlInputStreamSource {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        if (savedInstanceState == null) {
            loadFragment();
        }
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        Log.d(MainActivity.class.getSimpleName(), "New intent: " + intent);
        setIntent(intent);
        loadFragment();
    }

    @Override
    public InputStream getXmlInputStream() throws IOException {
        Uri data = getIntent().getData();
        Log.d(MainActivity.class.getSimpleName(), "Data uri: " + data);
        return (data == null) ? getSampleInputStream() : new FileInputStream(
                new File(data.getPath()));
    }

    private void loadFragment() {
        XmlDocumentFragment xmlDocumentFragment = new XmlDocumentFragment();
        getSupportFragmentManager().beginTransaction().replace(R.id.container,
                                                               xmlDocumentFragment).commit();
    }

    private InputStream getSampleInputStream() throws IOException {
        return getResources().getAssets().open("sample.xml");
    }

}

File: app/src/main/java/com/rubyhuntersky/angleedit/app/MainActivity.kt
package com.rubyhuntersky.angleedit.app

import android.content.Context
import android.net.Uri
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import com.rubyhuntersky.angleedit.app.MainActivityMessage.SetError
import com.rubyhuntersky.angleedit.app.MainActivityMessage.SetSource
import com.rubyhuntersky.angleedit.app.tools.*
import kotlinx.android.synthetic.main.cell_source.view.*
import java.sql.Timestamp

class MainActivity : BaseActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        if (savedInstanceState == null) {
            display()
        }
    }

    fun update(message: MainActivityMessage) = when (message) {
        is SetSource -> {
            RecentSources.add(RecentSource(message.sourceUri, Timestamp(System.currentTimeMillis())))
            startActivity(XmlDocumentActivity.newIntent(this as Context, message.sourceUri))
            overridePendingTransition(R.anim.slide_in_right, R.anim.slide_out_left_fade_out)
        }
        is SetError -> {
            showError(message.place, message.throwable)
        }
    }

    private fun display() {
        val nextFragment = RecentSourcesFragment()
        supportFragmentManager
                .beginTransaction()
                .replace(R.id.container, nextFragment, ACTIVE_FRAGMENT)
                .commit()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.activity_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.action_open) {
            alertDialog(this) {
                val view = layoutInflater.inflate(R.layout.cell_source, null)
                titleStringId = R.string.change_source
                bodyView = view
                buttons(AlertDialogButton.Negative("Cancel") {},
                        AlertDialogButton.Positive("Done") {
                            val sourceUri = Uri.parse(view.urlEditText.text.toString().trim())
                            update(SetSource(sourceUri))
                        })
            }.show()
        }
        return super.onOptionsItemSelected(item)
    }

    companion object {
        val TAG: String = MainActivity::class.java.simpleName
        val ACTIVE_FRAGMENT = "active-fragment"
    }
}
